features/telemetry/adapters/netflow/src/main/java/org/opennms/netmgt/telemetry/adapters/netflow/Netflow5Adapter.java

```diff
@@ -53,27 +61,86 @@ public void handleMessageLog(TelemetryMessageLogDTO messageLog) {
-                }
-                if (flowPacket.getCount() == 0) {
-                    LOG.warn("Received packet has no content. Skipping flow packet.");
-                    continue;
-                }
+            final NetflowPacket flowPacket = parseMessage(eachMessage);
+            if (flowPacket != null) {
+                LOG.debug("Flow packet received: {}. Try persisting.", flowPacket);
+                persist(flowPacket, messageLog);
+            }
+        }
+    }
+    
+    private NetflowPacket parseMessage(TelemetryMessageDTO messageDTO) {
+        // Create NetflowPacket which delegates all calls to the byte array
+        try {
+            final NetflowPacket flowPacket = new NetflowPacket(messageDTO.getBytes());
+            if (flowPacket.getVersion() != NetflowPacket.VERSION) {
+                LOG.warn("Invalid Version. Expected {}, received {}. Skipping flow packet.", NetflowPacket.VERSION, flowPacket.getVersion());
+            } else if (flowPacket.getCount() == 0) {
+                LOG.warn("Received packet has no content. Skipping flow packet.");
+            } else if (!flowPacket.isValid()) {
                 // TODO MVR an invalid packet is skipped for now, but we may want to persist it anyways
-                if (!flowPacket.isValid()) {
-                    LOG.warn("Received packet is not valid. Skipping flow packet.");
-                    continue;
-                }
-                LOG.debug("Flow packet received: {}", flowPacket);
-
-            } catch (Exception e) {
-                LOG.error("Received packet cannot be read.", e);
+                LOG.warn("Received packet is not valid. Skipping flow packet.");
+            } else {
+                return flowPacket;
             }
+        } catch (Throwable e) {
+            LOG.error("Received packet cannot be read.", e);
+        }
+        return null;
+    }
+    
+    private void persist(NetflowPacket netflowPacket, TelemetryMessageLogDTO messageLog) {
+        final List<NetflowDocument> flowDocuments = convert(netflowPacket, messageLog);
+        try {
+            provider.getFlowRepository().save(flowDocuments);
+        } catch (Throwable e) {
+            // TODO MVR deal with this accordingly
+            LOG.error("Error", e);
         }
     }
+
+    private List<NetflowDocument> convert(NetflowPacket netflowPacket, TelemetryMessageLogDTO messageLog) {
+        return netflowPacket.getRecords().stream()
+                .map(record -> {
+                    final NetflowDocument document = new NetflowDocument();
+
+                    // META
+                    document.setExporterAddress(InetAddressUtils.toIpAddrString(messageLog.getSourceAddress()));
+                    document.setLocation(messageLog.getLocation());
+                    document.setFlowType(FlowType.NETFLOW_5);
+                    document.setTimestamp(netflowPacket.getUnixSecs() * 1000L + netflowPacket.getUnixNSecs() / 1000L / 1000L);
```

Comment:
Can we also include some basic meta data related to the node? Perhaps we can include the Node ID, FS/FID if available and the list categories associated with the node.

We could also address this in a separate issue.

Reply:
Yes, the idea is to enrich each flow with such data. This however was left out on purpose in this issue.